Add Page Lifecycle API article

philipwalton opened the pull request:
**Target Live Date:** 2018-07-24

- [ ] This has been reviewed and approved by (@jpmedley)
- [x] I have run `npm test` locally and all tests pass.
- [x] I have added the appropriate `section-updates` label.
- [x] I've staged the site and manually verified that my content displays correctly.

**CC:** @petele

philipwalton commented:
/cc @shubhie

petele commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
````diff
@@ -0,0 +1,962 @@
+}, {capture: true});
+```
+
+<aside class="warning">
````
Tip: You can just do:
```
Warning: Some text here, you don't need to wrap this in an aside. DevSite does that for you automatically.
```

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+3. Older versions of Internet Explorer (10 and below) do not implement the
+   `visibilitychange` event.
+4. The ordering of when `pagehide` and `visibilitychange` events are dispatched
+   for unloading pages has
```
Change to 'The  dispatch order of the `pagehide` and `visibilitychange` events has changed.

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+4. The ordering of when `pagehide` and `visibilitychange` events are dispatched
+   for unloading pages has
+   [changed](https://github.com/w3c/page-visibility/issues/39). Previously
+   browsers would dispatch `visibilitychange` after `pagehide` if the page was
```
Change to 'browsers dispatched `pagehide` first if...'

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+   [changed](https://github.com/w3c/page-visibility/issues/39). Previously
+   browsers would dispatch `visibilitychange` after `pagehide` if the page was
+   visible while being unloaded. New Chrome versions will dispatch `pagehide`
+   after `visibilitychange`, regardless of the document's visibility state prior
```
New Chrome versions dispatch `visibilitychange` first regardless...

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+   browsers would dispatch `visibilitychange` after `pagehide` if the page was
+   visible while being unloaded. New Chrome versions will dispatch `pagehide`
+   after `visibilitychange`, regardless of the document's visibility state prior
+   to the page being unloaded.
```
Change 'prior to the page being unloaded' to 'before the page was unloaded'.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+inconsistencies and focus solely on following [lifecycle best practices
+](#developer-recommendations-for-each-state), we've released
+[PageLifecycle.js](https://github.com/GoogleChromeLabs/page-lifecycle), a
+JavaScript library for observing Page Lifecycle API state changes.
```
Is this not a polyfill?

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+[PageLifecycle.js](https://github.com/GoogleChromeLabs/page-lifecycle)
+[normalizes cross-browser differences in event firing order so that state change
+[always occur exactly as outlined in the chart and tables in this article (and
+[do so consistently in all browsers).
```
There are extra square brackets on these lines.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+JavaScript library for observing Page Lifecycle API state changes.
+
+[PageLifecycle.js](https://github.com/GoogleChromeLabs/page-lifecycle)
+[normalizes cross-browser differences in event firing order so that state change
```
'change' to 'changes'

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+should (and should not) be doing depends largely on what state your page is in.
+
+For example, it clearly doesn't make sense to be updating the inline styles of
+DOM elements if your page is in the hidden state. While this example is pretty
```
Change to 'For example, it doesn't make sense to update the inline styles of DOM elements while the page's state is `hidden`.'

I didn't flag them all, but states and events should be have backticks around them. Please fix throughout.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      to <a href="/web/updates/2015/08/using-requestidlecallback">
+      idle periods</a> or <a
+      href="/web/fundamentals/performance/rendering/optimize-javascript-execution#reduce_complexity_or_use_web_workers">
+      offloaded to a web worker</a>.</p>
```
What about `requestIdleCallback()`?

philipwalton replied:
The first link here points to that.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      they can still see it. This means UI updates and animations should still
+      be smooth, but the timing of when these updates occur is less critical.</p>
+      <p>When the page changes from <em>active</em> to <em>passive</em>, it's a
+      good time to persist unsaved application state.</p>
```
How about 'The `passive` state is a good time to persist unsaved application state.

philipwalton replied:
I think it's better to recommend it during the change. In general we don't want to encourage extra work being done in the passive state.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      mobile, as users can close tabs or the browser app itself, and the
+      pagehide and unload events are not fired in those cases).</p>
+      <p>This means you should treat the hidden state as the likely end to the
+      user's session &mdash; including persisting any unsaved application state,
```
'including' cannot modify the verb 'treat' in this sentences. Start a new sentence with 'In other words, persist any ...'

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      <p>This means you should treat the hidden state as the likely end to the
+      user's session &mdash; including persisting any unsaved application state,
+      and sending any unsent analytics data.</p>
+      <p>You should also stop making any UI updates (since they won't be seen
```
Delete 'any'.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+        <li>Release any held <a href="https://github.com/inexorabletash/web-locks">
+          Web Locks</a>.</li>
+      </ul>
+      <p>You should also persist any dynamic view state (e.g. scroll position
```
'state' to 'states'

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+          Web Locks</a>.</li>
+      </ul>
+      <p>You should also persist any dynamic view state (e.g. scroll position
+      in an infinite list view) to sessionStorage (or
```
Backticks again.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      <p>You should also persist any dynamic view state (e.g. scroll position
+      in an infinite list view) to sessionStorage (or
+      <a href="#save-data-to-indexeddb-before-freezing">IndexedDB via
+      <code>commit()</code></a>) that you'd want restored if the page was
```
'was' to 'were'.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      in an infinite list view) to sessionStorage (or
+      <a href="#save-data-to-indexeddb-before-freezing">IndexedDB via
+      <code>commit()</code></a>) that you'd want restored if the page was
+      discarded and then reloaded later.</p>
```
Delete 'then'.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      <a href="#save-data-to-indexeddb-before-freezing">IndexedDB via
+      <code>commit()</code></a>) that you'd want restored if the page was
+      discarded and then reloaded later.</p>
+      <p>If the page then transitions from <em>frozen</em> back to
```
Delete 'then'.
Use &lt;code> instead of &lt;em>. Please fix throughout.

philipwalton replied:
The states aren't really code, nor technically part of the API, so I don't think `<code>` makes sense for them.

I can remove the `<em>` if you think they're not appropriate. I know some people mark up technical terms using `<i>`.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      discarded and then reloaded later.</p>
+      <p>If the page then transitions from <em>frozen</em> back to
+      <em>hidden</em>, you can restore any connections or restart any polling
+      your stopped when the page was initially frozen.</p>
```
Something went wrong with this sentence.

philipwalton replied:
Fixed.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+   </td>
+  </tr>
+  <tr>
+    <td><strong>Discarded</strong></td>
```
These should be in &lt;code> also. Please change throughout.

philipwalton replied:
Same comment as before. These are concepts, but they're not part of the API so I don't think it makes sense to use `<code>`.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+## Legacy lifecycle APIs to avoid {: #legacy-lifecycle-apis-to-avoid}
+
+### The unload event {: #the-unload-event}
```
Do NOT use backticks or &lt;code> in headings.

philipwalton replied:
I don't see where I'm doing that in this line.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+<aside>
+  <strong>Advice summary:</strong>
+  never use the <code>unload</code> event on modern browsers
```
'Never' should be upper case. Sentences needs a period.

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+switcher on mobile or closing the browser app from the app switcher.
+
+For this reason, it's always better to rely on the
+[visibilitychange](#event-visibilitychange) event to determine session ends, and
```
'session ends' to 'when a session ends'

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+<aside>
+  <strong>Advice summary:</strong>
+  never add a <code>beforeunload</code> listener unconditionally or use it as an
```
'Never' should start upper case.

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+legitimate uses of `beforeunload`. For instance, when you want to warn the user
+that they have unsaved changes they'll lose if they continue unloading the page.
+
+Since there are valid reasons to use `beforeunload` but using it disallows pages
```
'disallows' to 'prevents'

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+Since there are valid reasons to use `beforeunload` but using it disallows pages
+from being added to the page navigation cache, it's recommended that you *only*
+add `beforeunload` listeners when a user has unsaved changes, and then you
```
Delete 'you'

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+DOMContentLoaded events don't signal a lifecycle state change, they're not
+relevant to this discussion.
+
+### If I can't run async APIs in the frozen or terminated states, how can I save data to IndexedDB? {: #save-data-to-indexeddb-before-freezing}
```
spell out asynchronous
check throughout

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+and callback-based APIs such as IndexedDB cannot be reliably used.
+
+In the future, we will [add a `commit()` method to
+`IDBTransaction`](https://github.com/w3c/IndexedDB/pull/242) objects, which will
```
PRs to a spec are sometimes hard to parse because I often have to expand hidden content to get context. Is it possible to get the change merged before this article is published so we can use a link to heading in the spec?

philipwalton replied:
Probably not since i'm wanting to publish this either tonight or tomorrow before the "new in Chrome 68" article goes live.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+give developers a way to perform what are effectively write-only transactions
+that don't require callbacks. In other words, if the developer is just writing
+data to IndexedDB and not performing a complex transaction consisting of reads
+and writes, the `commit()` method will be able to finish before tasks queues are
```
'tasks' to 'task'

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+that don't require callbacks. In other words, if the developer is just writing
+data to IndexedDB and not performing a complex transaction consisting of reads
+and writes, the `commit()` method will be able to finish before tasks queues are
+suspended (assuming the IndexedDB database is already open).
```
Do you mean 'still open'?

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+2. **Use IndexedDB via service worker:** service worker can store data in
+IndexedDB after the page has been terminated or discarded. In the `freeze` or
+`pagehide` event listener you can send data to a service worker via postMessage,
+and the service worker can handle saving the data.
```
These need to be bullets.

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+For code that needs to work today, however, developers have two options:
+
+1. **Use Session Storage**: Session Storage is synchronous and is persisted
```
Put the colon inside the bolding.

philipwalton replied:
Done.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+1. **Use Session Storage**: Session Storage is synchronous and is persisted
+across page discards.
+2. **Use IndexedDB via service worker:** service worker can store data in
```
Should be 'a service worker' in both instances.

philipwalton replied:
Updated to "your" and "a" respectively.

jpmedley commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+across page discards.
+2. **Use IndexedDB via service worker:** service worker can store data in
+IndexedDB after the page has been terminated or discarded. In the `freeze` or
+`pagehide` event listener you can send data to a service worker via postMessage,
```
postMessage to `postMessage()`

philipwalton replied:
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+<a href="/web/updates/images/2018/07/page-lifecycle-api-state-event-flow.png">
+  <img src="/web/updates/images/2018/07/page-lifecycle-api-state-event-flow.png"
+    alt="Page Lifecycle API state and event flow">
+</a>
```
In what way?

Here is the source if you want to make a copy: https://docs.google.com/drawings/d/176gNTdF6fTf4UFrzGjGClV6fSpkyOvqxSclglsn1c_Y/edit?usp=sharing

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+{% include "web/_shared/contributors/philipwalton.html" %}
+
+Modern browsers today will sometimes suspend pages or discard them entirely when
```
Hmmm, I like how saying "today" balances with saying "in the future" next.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+## Background {: #background}
+
+Application lifecycle is a key way that modern operating systems manage
+resources. On Android, iOS, and recently Windows, apps can be started and
```
What about "and recent Windows versions"?

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+* Introducing and standardizing the concept of lifecycle states on the web.
+* Defining new, system-initiated states that allow browsers to limit the
+  resources that can be consumed by inactive tabs.
+* Create new APIs and events that allow web developers to respond to transitions
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+* Defining new, system-initiated states that allow browsers to limit the
+  resources that can be consumed by inactive tabs.
+* Create new APIs and events that allow web developers to respond to transitions
+  to (and from) these new system-initiated states.
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+applications resilient to system interventions, and it allows browsers to more
+aggressively optimize system resources, ultimately benefiting all web users.
+
+This post will introduce the new Page Lifecycle features shipping in Chrome 68
```
Agreed, and updated.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+This post will introduce the new Page Lifecycle features shipping in Chrome 68
+and explore how they relate to all the existing web platform states and events.
+It will also give recommendations and best-practices for the types of work
+developers should (and should not) be doing in each state.
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+   <td>
+      <p>The document's visibilityState value has changed. This can happen when a
+      user navigates to a new page, switches tabs, closes a tab, minimizes or
+      closes the browser, or switches apps on mobile operating systems.</p>
```
No "a" because "systems" is plural.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+    <td>
+      <p>A session history entry is being traversed to.</p>
+      <p>This could be either a brand new page load or a page taken from the
+      page navigation cache. If the page was taken from the page navigation
```
It's a general term I made up because I didn't want to use the Webkit version (Page Cache) or the Firefox version (Back-Forward Cache or bfcache), but I agree it could probably use some clarity.

Maybe I'll add it to the FAQs and link to that entry any time I use it.

philipwalton replied:
Done.

jpmedley replied:
I won 't hold up because of this, but best practice is to give at least enough to make sense of the thing at the first occurrence. Then the 'more information' link is for more depth, not the basics.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      point.</p>
+      <aside class="warning">
+        <strong>Warning:</strong>
+        the beforeunload event should only be used to alert the user of unsaved
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+        <strong>Warning:</strong>
+        the beforeunload event should only be used to alert the user of unsaved
+        work. It should not be added unconditionally to the page, as doing so
+        can hurt performance in some cases. See the legacy APIs section for
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+        the beforeunload event should only be used to alert the user of unsaved
+        work. It should not be added unconditionally to the page, as doing so
+        can hurt performance in some cases. See the legacy APIs section for
+        more details.
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      <p>The page is being unloaded.</p>
+      <aside class="warning">
+        <strong>Warning:</strong>
+        using the unload event is not recommended because it's unreliable And
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+      <aside class="warning">
+        <strong>Warning:</strong>
+        using the unload event is not recommended because it's unreliable And
+        can hurt performance in some cases. See the legacy APIs section for
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+### New features added in Chrome 68 {: #new-features-added-in-chrome-68}
+
+The chart above shows two states that are system-initiated rather than user-
+initiated. Those states are [frozen](#state-frozen) and [discarded](#state-discarded).
```
Reworded.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+
+The chart above shows two states that are system-initiated rather than user-
+initiated. Those states are [frozen](#state-frozen) and [discarded](#state-discarded).
+As mentioned above, browsers today already do occasionally freeze and discard
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+## Observing Page Lifecycle states in code {: #observing-page-lifecycle-states-in-code}
+
+In the [active](#state-active), [passive](#state-passive), and [hidden](#state-hidden)
+states, it's possible to run JavaScript code that can determine the current
```
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
````diff
@@ -0,0 +1,962 @@
+```
+
+The [frozen](#state-frozen) and [terminated](#state-terminated) states, on the
+other hand, the state can only be detected in their respective event listener
````
Done.

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
````diff
@@ -0,0 +1,962 @@
+Lifecycle state changes with the following code.
+
+```
+// Sets the initial state using the `getState()` function (defined above).
````
Reworded to "stores".

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+2. It defines a function that accepts a next state and, if there's a change,
+   logs the state changes to the console.
+3. It adds capturing events listeners for all necessary lifecycle events, which
+   in turn call `logStateChange()`, passing in the next state.
```
Done, but "in tern calls" is correct here.

jpmedley replied:
Spelling. Duh!

philipwalton commented on src/content/en/updates/2018/07/page-lifecycle-api.md:
```diff
@@ -0,0 +1,962 @@
+1. It sets the initial state using the `getState()` function.
+2. It defines a function that accepts a next state and, if there's a change,
+   logs the state changes to the console.
+3. It adds capturing events listeners for all necessary lifecycle events, which
```
Done.

philipwalton commented:
Thanks @jpmedley for the detailed review! I think I've addressed all the issues. Can you PTAL so I can publish as soon as possible? I'm trying to get this out before @petele's article tomorrow (which links to this).

petele commented:
@jpmedley - most of the feedback has been addressed, and this needs to go out with my NIC post today. I'm going to merge, and anything outstanding can be addressed in a future PR.
